### PR TITLE
Enable WebGPU GQA local-window flash attention

### PR DESCRIPTION
**Pull request by feich-ms**
### Description

Complete WebGPU `GroupQueryAttention` local-window support in shared flash attention, including graph-captured decode.

Following the generic flash-prefill local-window support added by #32277, this change completes the split-reduce path and removes GQA's host-side sliding-window dispatch restriction. GQA now forwards `local_window_size` to flash attention whenever the configuration is otherwise supported, regardless of whether graph capture is enabled.

The split-reduce shader now:

- Reads the logical sequence length from GPU-resident `seqlens_k` during graph replay.
- Applies the local window independently for each query row, supporting single-token decode and short multi-token inputs.
- Emits neutral per-row metadata for KV tiles outside that query row's window.
- Avoids zero-sum normalization for fully excluded tiles.

Shared-KV (`kv_sequence_length == 0`) layers preserve their established no-window behavior because they reuse another layer's prepared KV cache. `smooth_softmax` and other unsupported flash-attention configurations remain independently gated.

Together with #32277, local-window GQA now uses the shared flash-attention dispatch cascade for:

- Eager and graph-captured decode through split-reduce flash attention.
- Short multi-token inputs through split-reduce flash attention.
- Prefill through generic flash attention.

Regression coverage includes:

- Packed-QKV rotary graph capture compared with eager WebGPU across the local-window boundary.
- Multi-token split-reduce with per-query-row window starts.
- Generic flash prefill with a local window.
- Existing shared-KV behavior.

### Motivation and Context

Gemma 4 uses both global-attention layers and local-attention layers with `local_window_size=512`.

WebGPU GQA previously selected between flash attention and non-flash sliding-window attention using a host-side sequence length. During graph capture, the logical sequence length is GPU-resident. If the graph was captured before reaching 512 tokens, it permanently recorded ordinary flash attention and continued attending to the full KV history after crossing the boundary.

This produced increasingly incorrect logits during long generation. Eager WebGPU remained coherent because it reevaluated the host-side path selection on every decoding step, but maintaining separate eager and captured policies left the implementations inconsistent and prevented local-window GQA prefill from using flash attention.

This change moves the remaining local-window behavior into shared split-reduce flash attention and relies on #32277 for generic flash prefill, making shader masking the source of truth instead of host-side kernel-family switching.

Observed before the fix:

| Step | Logit cosine | Maximum difference |
|---:|---:|---:|
| 300 | 1.000000 | 0.000 |
| 500 | 0.999993 | 0.250 |
| 600 | 0.975641 | 28.375 |
| 650 | 0.973550 | 42.453 |

Generation subsequently diverged and collapsed into repeated tokens.

After the fix:

| Step | Logit cosine | Maximum difference |
|---:|---:|---:|
| 300 | 1.000000 | 0.000 |
| 500 | 0.999997 | 0.125 |
| 600 | 0.999997 | 0.140625 |
| 650 | 0.999990 | 0.15625 |

Captured generation remains coherent and reaches EOS normally. The remaining small differences are expected from floating-point execution differences between attention paths.

### CODE REVIEW

File: onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc [hunk 1]
```diff
@@ -426,7 +431,7 @@ Status GroupQueryAttention::ComputeInternal(onnxruntime::webgpu::ComputeContext&
       // query points to packed QKV, K and V are nullptr since they're not needed
       return ApplyFlashAttention(query, nullptr, nullptr, attention_bias, output, past_key, present_key, past_value,
                                  present_value, parameters, context, seqlen_k, cos_cache, sin_cache, head_sink,
-                                 total_seqlen_tensor);
+                                 total_seqlen_tensor, nullptr, nullptr, 0, 0, nullptr, -1);
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc]: Addressed in d812c93b4c: the packed-QKV rotary static-cache early-return now forwards local_window_size_, so captured packed models use the same dynamic shader window.

File: onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template [hunk 2]
```diff
@@ -102,6 +102,24 @@ $MAIN {
   #else
   let total_sequence_length = global_total_sequence_length;
   #endif
+  let window_start = select(0u,
+                            total_sequence_length - uniforms.local_window_size,
+                            uniforms.local_window_size > 0u &&
+                                total_sequence_length > uniforms.local_window_size);
+  // Whole tiles outside the window must publish a zero softmax sum. Masking every
+  // element to the finite fp16 minimum would otherwise give the tile nonzero weight.
+  if (total_seq_offset >= total_sequence_length || total_seq_offset + tile_size <= window_start) {
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template:111]: Addressed in d812c93b4c: the replay length is now 129 with a 130-token cache and a 64-token window, which wholly excludes the first 64-token tile. The targeted test passes: GroupQueryAttentionTest.WebGPU_GraphCapture_ReplayCrossesLocalWindowBoundary.

File: onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template [hunk 3]
```diff
@@ -243,6 +261,11 @@ $MAIN {
         sum = q_element_t(-65504.0f);
       }
 #endif
+      let causal_sequence_length = past_sequence_length + q_idx + 1u;
+      if (uniforms.local_window_size > 0u &&
+          total_seq_offset + local_idx + uniforms.local_window_size < causal_sequence_length) {
+        sum = q_element_t(-65504.0f);
+      }
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template:267]: Addressed in 16a901babc: the local-window calculation now remains inside #if is_unidirectional, so non-causal Attention/MultiHeadAttention shader variants do not reference past_sequence_length. Rebuilding onnxruntime_provider_test regenerated and compiled all WebGPU variants successfully.

File: onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template (hunk 2 above)
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template:111]: Addressed in 16a901babc: GQA now forwards a positive flash window only for graph-captured single-token decode. Multi-token and shared-KV calls receive -1, preserving their existing behavior. The boundary regression still passes at replay length 129.

File: onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc [hunk 4]
```diff
@@ -385,11 +385,17 @@ Status GroupQueryAttention::ComputeInternal(onnxruntime::webgpu::ComputeContext&
 
   // Use a sliding window if the total sequence exceeds the window's length.
   bool use_sliding_window = (local_window_size_ != -1 && local_window_size_ < parameters.total_sequence_length_);
+  // During graph capture the logical sequence length lives on the GPU. Keep decode on
+  // flash attention and apply the local window dynamically from seqlens_k in the shader.
+  const bool use_dynamic_flash_window = context.IsGraphCaptureEnabled() &&
+                                        parameters.sequence_length_ == 1 &&
+                                        local_window_size_ != -1;
+  const int flash_local_window_size = use_dynamic_flash_window ? local_window_size_ : -1;
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc:394]: Addressed in ff12ae7287: use_dynamic_flash_window now explicitly excludes kv_empty, preserving the established shared-KV behavior. GroupQueryAttentionTest.WebGPU_SharedKV_SlidingWindow passes.

File: onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc [hunk 5]
```diff
@@ -426,7 +432,7 @@ Status GroupQueryAttention::ComputeInternal(onnxruntime::webgpu::ComputeContext&
       // query points to packed QKV, K and V are nullptr since they're not needed
       return ApplyFlashAttention(query, nullptr, nullptr, attention_bias, output, past_key, present_key, past_value,
                                  present_value, parameters, context, seqlen_k, cos_cache, sin_cache, head_sink,
-                                 total_seqlen_tensor);
+                                 total_seqlen_tensor, nullptr, nullptr, 0, 0, nullptr, flash_local_window_size);
```
[the remark this answers is not in the document]
Comment (feich-ms) [reply onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc:436]: Addressed in ff12ae7287: added WebGPU_GraphCapture_PackedRotaryCrossesLocalWindowBoundary using packed QKV, rotary caches, aliased static KV buffers, and GPU-resident lengths crossing the 64-token window. The new test and the separate-QKV boundary test both pass.

**Review by qjia7 — COMMENTED** [review]
## Review frame
- Problem/feature validity: Validated. WebGPU graph capture fixes the host-side GQA dispatch decision while `seqlens_k` and `total_sequence_length` remain GPU-resident and mutable across replay. A graph captured before the local-window boundary can therefore keep using unwindowed flash attention after the logical length crosses the boundary.
- Risk/scope: Deep. This is a graph-capture correctness fix, but `local_window_size` is an attention semantic shared by eager and captured execution, prefill and decode, and GQA and PagedAttention. The implementation also changes shared flash-attention APIs and shader masking/metadata.
- Direction gate: Fail. The PR adds a GQA-specific implementation for graph-captured single-token split-reduce while #32277 adds the complementary local-window implementation to the generic flash-prefill shader but leaves it disabled for GQA. The owner solution is to complete local-window support in both shared flash-attention tiers, then remove the host-side GQA dispatch restriction instead of retaining a permanent prefill optimization gap and different eager/captured decode paths.

## Confirmed findings
### D1: Implement local-window semantics at the shared flash-attention boundary
`onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc:388-400`

Graph capture is decode-only, so restricting this new split-reduce mask to captured `sequence_length == 1` addresses the reported correctness failure. It still leaves local-window behavior split across implementations: eager GQA falls back to `ApplyAttention`, captured decode uses the new split-reduce mask, and GQA prefill cannot use the existing flash-attention optimization whenever `local_window_size` is active.

#32277 already implements the complementary half in the shared generic `FlashAttentionProgram`: it adds a per-query causal `local_window_size` mask to `flash_attention.wgsl.template`. However, it does not enable that support for GQA, and it explicitly routes every local-window call away from split-reduce with:

```cpp
const bool use_split_reduce = parameters.sequence_length_ < 32 && !has_local_window;
```

As a result, landing the two PRs independently would leave two partial policies in the same `ApplyFlashAttention` API: #32277 implements local windows in the generic prefill shader for PagedAttention, while this PR enables them only for graph-captured single-token GQA in split-reduce. GQA prefill would still be excluded from the optimization, eager GQA decode would still use the non-flash path, and the duplicate API edits would need reconciliation.

Please consolidate the work at the shared flash-attention boundary:

1. Reuse #32277's per-query local-window implementation in `FlashAttentionProgram` for the prefill tier.
2. Extend the split-reduce shader here to support local windows for every supported `m_tile`, emitting neutral metadata per query row when a tile is fully outside that row's window.
3. Once both tiers support the semantic, remove GQA's `!use_sliding_window` flash-selection restriction and pass `local_window_size_` through both GQA flash call sites, while preserving the established `kv_empty` shared-KV behavior.
4. Keep `use_smooth_softmax_` and genuinely unsupported paged/direct combinations gated independently from local-window support.

This makes the shader the single source of truth for the attention mask, allows eager and captured GQA decode to use the same dispatch cascade, enables flash-attention optimization for GQA prefill, and avoids a graph-capture-only policy that must be kept synchronized with #32277's prefill implementation.

## Clarifications
None.

## Test coverage
The current packed-rotary regression exercises captured single-token split-reduce, including partial-tile masking and a fully excluded leading tile. The consolidated implementation should retain captured-versus-eager decode coverage for split-reduce and add eager flash-versus-reference coverage for GQA prefill. It should retain packed-rotary coverage and add separate-Q/K/V coverage so both GQA call sites are exercised. Multi-token split-reduce coverage should use query rows with different window starts to validate per-row neutral metadata.

WGSL template checks and macOS WebGPU build-and-test jobs passed for the current head. The failing `web_Release / build_onnxruntime_web` job was a self-hosted-runner communication loss after its WebGPU EP test step passed, so it is not attributable to this diff.

## Verdict
The graph-capture correctness problem is valid, but D1 blocks the current direction. Local-window masking should be completed in both shared flash-attention tiers by combining this split-reduce work with #32277's generic prefill work, after which GQA can remove the host-side sliding-window exclusion and use one dispatch policy for eager and captured execution. There are no clarification requests or cleanup-only items.

**Review by qjia7 — COMMENTED** [review]
## Review frame
- Problem/feature validity: Validated. During WebGPU graph capture, GQA's logical sequence lengths remain GPU-resident and mutable across replay, so selecting an unwindowed flash path from the host-visible length produces incorrect decode after crossing `local_window_size`.
- Risk/scope: Deep. Current head `29abe189ed7df3805a3b9c2fbd19bbaa3001cb55` rebases onto the merged #32277 prefill implementation and changes shared flash-attention dispatch, split-reduce masking and metadata, eager GQA behavior, graph-captured decode, and performance for long histories.
- Direction gate: Pass. The prior D1 is resolved: GQA no longer excludes sliding-window flash attention, forwards `local_window_size_` through both flash call sites, preserves `kv_empty` shared-KV behavior by passing `-1`, reuses #32277's generic prefill mask, and implements per-query local-window bounds in split-reduce for every `m_tile`.

## Confirmed findings
### Q1: Restore the whole-window tile early exit for split-reduce
`onnxruntime/contrib_ops/webgpu/bert/flash_attention_decode_qkv.wgsl.template:110-123`

The earlier decode-only implementation returned before loading K/V when a complete tile was older than the local window. The generalized implementation removed that part of the condition and now exits early only for tiles beyond `total_sequence_length`. For a one-token decode at total length 32K with `local_window_size=512`, split-reduce therefore still dispatches roughly 500 KV-tile workgroups per head; about 492 of them load K/V, compute the full QK dot products and V products, and only then produce `l_sum=0` because the tile is outside the window. This makes the new local-window flash decode compute and read memory proportional to the full history rather than the configured window.

Please retain the per-query masking but add a safe workgroup-level rejection for tiles excluded by every query row in the workgroup. For causal `m_tile > 1`, the first valid query has the earliest window start, so a tile satisfying `total_seq_offset + tile_size <= first_query_window_start` can publish `(-FLT_MAX, 0)` metadata for each row and return before the K/V loads and dot products. This restores the original decode optimization without repeating its final-query cutoff bug.

## Clarifications
None.

## Test coverage
The prior coverage requests are resolved:

- `LocalWindowMultiToken_WebGPU` exercises the separate-Q/K/V call path and per-row masking in split-reduce with `m_tile=2`.
- `LocalWindowPrefill_WebGPU` exercises the separate-Q/K/V call path and GQA local-window masking through the generic 32-token prefill shader.
- `WebGPU_GraphCapture_PackedRotaryLocalWindow` compares captured and eager packed-rotary single-token decode across the 64-token boundary and includes fully excluded historical tiles.

All three local-window tests ran and passed in the macOS-arm64 Debug WebGPU job. WGSL template tests and all WebGPU build-and-test jobs are green. The failing Windows CUDA test job is unrelated: its failures are CUDA `test_paged_attention_int4.py` XQA kernel-selection assertions and do not execute the changed WebGPU paths.

Q1 is a structural performance issue rather than a numerical coverage gap; a focused output test cannot distinguish the rejected workgroups, but the existing captured/eager test protects the neutral-metadata result after restoring the early exit.

## Verdict
The graph-capture correctness problem is valid and the revised shared FlashAttention design is appropriate. The prior D1 design blocker is resolved. Q1 is the remaining actionable performance issue: fully window-excluded split-reduce tiles should return before K/V loads and QK/V computation. There are no clarification requests or cleanup-only items.

**Comment by feich-ms** [comment]
Addressed qjia7 Q1 in commit `9d62a3c065`: split-reduce now computes the earliest query row window start for each workgroup and returns before K/V loads when a tile is excluded by every row. The per-query masking remains unchanged. This is safe for all `m_tile` values because `q_base` is the earliest query in the workgroup. Validation: `onnxruntime_provider_test` rebuilt successfully and all 12 relevant local-window, graph-capture, shared-KV, 4-bit, and 8-bit indirect-dispatch tests passed.

File: onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc [hunk 6]
```diff
@@ -388,13 +388,9 @@ Status GroupQueryAttention::ComputeInternal(onnxruntime::webgpu::ComputeContext&
   Tensor qRotary;
   Tensor kRotary;
 
-  // Use a sliding window if the total sequence exceeds the window's length.
-  bool use_sliding_window = (local_window_size_ != -1 && local_window_size_ < parameters.total_sequence_length_);
+  const int flash_local_window_size = kv_empty ? -1 : local_window_size_;
```
Comment (qjia7) [inline onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc]:
```suggestion
  const int local_window_size = kv_empty ? -1 : local_window_size_;
```

**Review by qjia7 — APPROVED** [review]
